File: src/main/java/pl/miloszgilga/command/music/ShowQueueCmd.java
/*
 * Copyright (c) 2023 by MILOSZ GILGA <http://miloszgilga.pl>
 *
 * File name: ShowQueueCmd.java
 * Last modified: 16/05/2023, 18:52
 * Project name: jwizard-discord-bot
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this
 * file except in compliance with the License. You may obtain a copy of the License at
 *
 *     <http://www.apache.org/license/LICENSE-2.0>
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the license.
 */

package pl.miloszgilga.command.music;

import com.jagrosh.jdautilities.menu.Paginator;
import net.dv8tion.jda.api.entities.MessageEmbed;
import com.sedmelluq.discord.lavaplayer.track.AudioTrack;

import java.util.List;
import java.util.Queue;
import java.util.Objects;
import java.util.ArrayList;

import pl.miloszgilga.BotCommand;
import pl.miloszgilga.audioplayer.*;
import pl.miloszgilga.misc.Utilities;
import pl.miloszgilga.locale.ResLocaleSet;
import pl.miloszgilga.dto.QueueEmbedContent;
import pl.miloszgilga.dto.CommandEventWrapper;
import pl.miloszgilga.embed.EmbedMessageBuilder;
import pl.miloszgilga.embed.EmbedPaginationBuilder;
import pl.miloszgilga.command.AbstractMusicCommand;
import pl.miloszgilga.cacheable.CacheableCommandStateDao;
import pl.miloszgilga.core.remote.RemotePropertyHandler;
import pl.miloszgilga.core.configuration.BotConfiguration;
import pl.miloszgilga.core.loader.JDAInjectableCommandLazyService;

import static pl.miloszgilga.exception.AudioPlayerException.TrackQueueIsEmptyException;

////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

@JDAInjectableCommandLazyService
public class ShowQueueCmd extends AbstractMusicCommand {

    private final EmbedPaginationBuilder pagination;

    ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

    ShowQueueCmd(
        BotConfiguration config, EmbedPaginationBuilder pagination, PlayerManager playerManager,
        EmbedMessageBuilder embedBuilder, RemotePropertyHandler handler, CacheableCommandStateDao cacheableCommandStateDao
    ) {
        super(BotCommand.QUEUE, config, playerManager, embedBuilder, handler, cacheableCommandStateDao);
        this.pagination = pagination;
        super.onSameChannelWithBot = true;
    }

    ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

    @Override
    protected void doExecuteMusicCommand(CommandEventWrapper event) {
        final MusicManager musicManager = playerManager.getMusicManager(event);
        final Queue<AudioQueueExtendedInfo> tracks = musicManager.getQueue();
        if (tracks.isEmpty()) {
            throw new TrackQueueIsEmptyException(config, event);
        }
        final List<String> pageableTracks = new ArrayList<>();
        int i = 0;
        for (final AudioQueueExtendedInfo aTrack : tracks) {
            final AudioTrack track = aTrack.audioTrack();
            pageableTracks.add(Utilities.getRichPageableTrackInfo(++i, track));
        }
        String leftToNextTrack = "-";
        final ExtendedAudioTrackInfo currentTrack = playerManager.getCurrentPlayingTrack(event);
        if (!Objects.isNull(currentTrack)) {
            leftToNextTrack = Utilities.convertMilisToDate(currentTrack.getApproxTime());
        }
        final long durationMilis = tracks.stream().mapToLong(t -> t.audioTrack().getDuration()).sum();

        final QueueEmbedContent content = new QueueEmbedContent(
            String.valueOf(tracks.size()),
            Utilities.convertMilisToDate(durationMilis),
            leftToNextTrack,
            Utilities.convertMilisToDate(musicManager.getActions().getAverageTrackDuration()),
            musicManager.isInfinitePlaylistActive() ? ResLocaleSet.TURN_ON_MESS : ResLocaleSet.TURN_OFF_MESS
        );
        final MessageEmbed messageEmbed = embedBuilder.createQueueInfoMessage(content, event.getGuild());
        final Paginator paginator = pagination.createDefaultPaginator(pageableTracks);

        event.appendEmbedMessage(messageEmbed, () -> paginator.display(event.getTextChannel()));
    }
}

File: src/main/kotlin/pl/jwizard/core/api/music/ShowQueueCmd.kt
/*
 * Copyright (c) 2024 by JWizard
 * Originally developed by Miłosz Gilga <https://miloszgilga.pl>
 */
package pl.jwizard.core.api.music

import pl.jwizard.core.api.AbstractMusicCmd
import pl.jwizard.core.audio.ExtendedAudioTrackInfo
import pl.jwizard.core.audio.player.PlayerManagerFacade
import pl.jwizard.core.bot.BotConfiguration
import pl.jwizard.core.command.BotCommand
import pl.jwizard.core.command.CompoundCommandEvent
import pl.jwizard.core.command.embed.CustomEmbedBuilder
import pl.jwizard.core.command.embed.EmbedColor
import pl.jwizard.core.command.reflect.CommandListenerBean
import pl.jwizard.core.exception.AudioPlayerException
import pl.jwizard.core.i18n.I18nMiscLocale
import pl.jwizard.core.util.DateUtils
import pl.jwizard.core.util.Formatter
import net.dv8tion.jda.api.entities.MessageEmbed

@CommandListenerBean(id = BotCommand.QUEUE)
class ShowQueueCmd(
	botConfiguration: BotConfiguration,
	playerManagerFacade: PlayerManagerFacade
) : AbstractMusicCmd(
	botConfiguration,
	playerManagerFacade
) {
	init {
		onSameChannelWithBot = true
	}

	override fun executeMusicCmd(event: CompoundCommandEvent) {
		val musicManager = playerManagerFacade.findMusicManager(event)
		if (musicManager.queue.isEmpty()) {
			throw AudioPlayerException.TrackQueueIsEmptyException(event)
		}
		val pageableTracks = musicManager.queue
			.mapIndexed { index, track -> Formatter.createRichPageableTrackInfo(index, track) }

		val currentTrack = playerManagerFacade.findMusicManager(event).audioPlayer.playingTrack
		var leftToNextTrack = "-"
		if (currentTrack != null) {
			leftToNextTrack = DateUtils.convertMilisToDTF(ExtendedAudioTrackInfo(currentTrack).approximateTime)
		}
		val tracksPaginator = createDefaultPaginator(pageableTracks)
		val messageEmbed = createQueueMessage(
			event,
			queueSize = musicManager.queue.size,
			queueMaxDuration = musicManager.queue.sumOf { it.duration },
			timeToNextTrack = leftToNextTrack,
			averageTrackDuration = musicManager.actions.getAverageTracksDuration(),
			repeatingState = musicManager.actions.infiniteRepeating,
		)
		event.appendEmbedMessage(messageEmbed) { tracksPaginator.display(event.textChannel) }
	}

	private fun createQueueMessage(
		event: CompoundCommandEvent,
		queueSize: Int,
		queueMaxDuration: Long,
		timeToNextTrack: String,
		averageTrackDuration: Long,
		repeatingState: Boolean,
	): MessageEmbed = CustomEmbedBuilder(event, botConfiguration)
		.appendKeyValueField(I18nMiscLocale.ALL_TRACKS_IN_QUEUE_COUNT, queueSize)
		.addSpace()
		.appendKeyValueField(
			I18nMiscLocale.ALL_TRACKS_IN_QUEUE_DURATION,
			DateUtils.convertMilisToDTF(queueMaxDuration)
		)
		.appendKeyValueField(I18nMiscLocale.APPROX_TO_NEXT_TRACK_FROM_QUEUE, timeToNextTrack)
		.addSpace()
		.appendKeyValueField(
			I18nMiscLocale.PLAYLIST_AVERAGE_TRACK_DURATION,
			DateUtils.convertMilisToDTF(averageTrackDuration)
		)
		.appendKeyValueField(
			I18nMiscLocale.PLAYLIST_REPEATING_MODE,
			i18nService.getMessage(Formatter.toStateTag(repeatingState), event.guildId)
		)
		.addColor(EmbedColor.WHITE)
		.build()
}
